### PR TITLE
[wasm] Improvements to startup performance of mono_wasm_get_assembly_exports

### PR DESCRIPTION
**Pull request by kg**
Generated JSImport/JSExport initializer does an `Environment.Version` check to see if we're on NET7. That check is tremendously expensive, because it fetches a CustomAttribute instance off of something in corlib, then parses the version string (parsing version strings initializes a bunch of stuff). This PR removes that check and adds a linker dependency to keep the registration method alive so it can be called when needed.

We will also need to document that DLLs (in nugets?) generated by this new version of the generator won't work on NET7 runtimes, since the relevant logic to perform automatic registration on that version.

### CODE REVIEW

**Comment by kg** [comment]
Requesting review to get feedback. I hope someone knows trimming better than me and can suggest how to fix this. Put a dependency attribute of some kind on the generated module initializer, maybe?

File: src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSExportGenerator.cs [hunk 1]
```diff
@@ -253,8 +254,8 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                             }))
                             .WithBody(Block(
                                 IfStatement(BinaryExpression(SyntaxKind.EqualsExpression,
-                                    IdentifierName("Environment.Version.Major"),
-                                    LiteralExpression(SyntaxKind.NumericLiteralExpression, Literal(7))),
+                                    IdentifierName("System.Runtime.InteropServices.JavaScript.JSHost.GlobalThis"),
```
Comment (pavelsavara) [inline src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSExportGenerator.cs:257]: JSHost.GlobalThis is not cheap to evaluate. As I said before, I think we could drop the Net7 support and replace it with something else which will keep the initializer alive until explicitly called by JS.

Comment (kg) [reply src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSExportGenerator.cs:257]: Yes, that is why this is no-merge. I'm not sure what to use. Stopwatch.GetTimestamp is cheaper but ideally we could just protect the target method from trimming with an attribute somehow.

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.cs [hunk 2]
```diff
@@ -333,7 +333,7 @@ public static (string assemblyName, string nameSpace, string shortClassName, str
 
             var nameSpace = "";
             var shortClassName = className;
-            var idx = fqn.LastIndexOf(".");
+            var idx = fqn.LastIndexOf('.');
```
Comment (pavelsavara) [inline src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.cs:336]: maybe we need to use `StringComparison.Ordinal` for all `IndexOf` here ? Eventually you can move this to separate PR

**Comment by pavelsavara** [comment]
Let's rename `__Net7SelfInit_` to something like `__DontTrimRegister_`. Also please update description on top of this PR.

**Comment by pavelsavara** [comment]
Here we are removing support for 3rd party library/nuget generated by Net9 SDK from being able to run on Net7 runtime.
I think it's OK. @lewing thoughts ?

**Review by pavelsavara — APPROVED** [review]

**Comment by maraf** [comment]
> Here we are removing support for 3rd party library/nuget generated by Net9 SDK from being able to run on Net7 runtime. I think it's OK. @lewing thoughts ?

.NET 7 is EOL on May 14, 2024. Removing support seems reasonable to me

**Comment by radekdoulik** [comment]
image
works nicely
